### ISSUE TITLE
CiviDiscount "Codes Redeemed" data lost during merge procedure

### ISSUE THREAD

**Issue by abeconway** [dlobo/org.civicrm.module.cividiscount#136]
When you merge 2 duplicate individual records (using "Find and Merge Duplicate Contact Records under the Contacts menu) together that both contain "Codes Redeemed", that CiviDiscount data is completely lost/deleted and NOT merged into the final record.

Here is the merge screen, notice there is nothing about merging the CiviDiscount data:
![cividiscount-mergescreen](https://cloud.githubusercontent.com/assets/16328640/11908709/6884f494-a593-11e5-9fa0-a5ac74ba6cfa.gif)

Here is what MySQL says BEFORE the merge:
mysql> select id,item_id,contact_id,contribution_id from cividiscount_track where contact_id IN (66855,2492);
+------+---------+------------+-----------------+
| id   | item_id | contact_id | contribution_id |
+------+---------+------------+-----------------+
|  157 |      28 |       2492 |           16023 |
|  757 |      73 |       2492 |           17309 |
| 1501 |      73 |      66855 |           18229 |
+------+---------+------------+-----------------+
3 rows in set (0.00 sec)

Here is what MySQL says AFTER the merge:
mysql> select id,item_id,contact_id,contribution_id from cividiscount_track where contact_id IN (66855,2492);
+-----+---------+------------+-----------------+
| id  | item_id | contact_id | contribution_id |
+-----+---------+------------+-----------------+
| 157 |      28 |       2492 |           16023 |
| 757 |      73 |       2492 |           17309 |
+-----+---------+------------+-----------------+
2 rows in set (0.00 sec)

If you check the Discount Code usage in the CiviCRM Discount Codes Page, it also has no record of the usage, example: Before the merge it would say 2/10 codes used, but AFTER the merge is reports 1/10 codes used.

CiviCRM 4.4.20
CiviDiscount 2.6
Drupal 6.37

**Comment by eileenmcnaughton** [comment]
Note to whoever looks to fix this - you need to start by looking at 

https://issues.civicrm.org/jira/browse/CRM-17454 & 
https://wiki.civicrm.org/confluence/display/CRMDOC/hook_civicrm_entityTypes

After 4.4 (I think) someone added a way to avoid this by querying the information schema. The downside is that the query is incredibly slow and locks your whole server and can lead to corruption.

In 4.7 (and Fuzion 4.6 repo) the method has changed to be based on metadata from CiviCRM rather than querying the table. The side-effect is that you need to ensure the metadata is declared for the extension

**Comment by twomice** [comment]
@eileenmcnaughton This ticket might be a convenient place to address the issue you mention (CRM-17454), but I don't think the two are directly related. Are you saying that this issue is blocked by CRM-17454?

**Comment by eileenmcnaughton** [comment]
From when CRM-17454 is applied the merge bug should be fixed IF the mentioned hook is used, reason being that the choice about what to merge is based on declared entities

**Comment by eileenmcnaughton** [comment]
The Item entity needs to be added here https://github.com/dlobo/org.civicrm.module.cividiscount/blob/master/cividiscount.php#L1163 to make this work
